### ISSUE TITLE
Feature Request: Allow client to dictate solely dictate transtype

### ISSUE THREAD

**Issue by oviano** [Haivision/srt#655]
My app currently uses the same TCP port number for a control connection (needs to be reliable eg file transtype) and a stream connection (live transtype).

Trying to get things working with SRT it seems that the server listen transtype has to match the client connect transtype.

I would like the client to be able to dictate the server’s transtype (ie match the client) either by default or at least have this as an option you can set on the server (SRTT_AUTO?)

At the moment I would need to have the server listen on a different port in order to have concurrent file + live connections.

**Comment by ethouris** [comment]
We had an idea for that, but it requires adding a larger feature, which would allow that several options be set on a socket that normally can't be altered after connecting (so they currently can't be also altered on an accepted socket).

**Comment by oviano** [comment]
Fair enough I’ll work around that for now then!

**Comment by maxsharabayko** [comment]
Likely possible now (to be confirmed) using:
1. SRT v1.4.3 or higher,
2. [the listener callback](https://github.com/Haivision/srt/blob/v1.4.4/docs/API/API-functions.md#srt_listen_callback),
3. PR #1652 (available since SRT v1.4.3).

**Comment by maxsharabayko** [comment]
✔️ **Confirmed**: possible with SRT v1.4.3 or higher.

The client can signal it wants to connect in file transfer mode following the [Access Control Guidelines](https://github.com/Haivision/srt/blob/master/docs/features/access-control.md) (`t=file`).
If the server allows this kind of connection, it can set the `SRTO_TRANSTYPE` accordingly from the listener callback.
An example test case based on the `TEST(Core, ListenCallback)` (see `SrtTestListenCallback`):

```c++
srt_listen_callback_fn SrtTestListenCallback;

TEST(Core, ListenCallback) {
    using namespace std;

    ASSERT_EQ(srt_startup(), 0);
    // Create server on 127.0.0.1:5555
    const SRTSOCKET server_sock = srt_create_socket();
    ASSERT_GT(server_sock, 0);    // socket_id should be > 0

    sockaddr_in bind_sa;
    memset(&bind_sa, 0, sizeof bind_sa);
    bind_sa.sin_family = AF_INET;
    ASSERT_EQ(inet_pton(AF_INET, "127.0.0.1", &bind_sa.sin_addr), 1);
    bind_sa.sin_port = htons(5555);

    ASSERT_NE(srt_bind(server_sock, (sockaddr*)&bind_sa, sizeof bind_sa), -1);
    ASSERT_NE(srt_listen(server_sock, 5), -1);
    (void)srt_listen_callback(server_sock, &SrtTestListenCallback, NULL);

    // Create client to connect to the above server
    SRTSOCKET client_sock;
    sockaddr_in sa;
    memset(&sa, 0, sizeof sa);
    sa.sin_family = AF_INET;
    sa.sin_port = htons(5555);
    ASSERT_EQ(inet_pton(AF_INET, "127.0.0.1", &sa.sin_addr), 1);
    sockaddr* psa = (sockaddr*)&sa;

    cerr << "TEST: Connect with TRANSTYPE file\n";
    client_sock = srt_create_socket();
    ASSERT_GT(client_sock, 0);    // socket_id should be > 0

    int optval = SRTT_FILE;
    ASSERT_NE(srt_setsockflag(client_sock, SRTO_TRANSTYPE, &optval, sizeof optval), -1);

    // EXPECTED RESULT: connection rejected
    EXPECT_NE(srt_connect(client_sock, psa, sizeof sa), SRT_ERROR);

    SRTSOCKET acceptsock = srt_accept(server_sock, NULL, NULL);
    cerr << "Accept result " << acceptsock << '\n';

    // Close the socket
    EXPECT_EQ(srt_close(client_sock), SRT_SUCCESS);

    (void)srt_cleanup();
}

int SrtTestListenCallback(void* opaq, SRTSOCKET ns SRT_ATR_UNUSED, int hsversion, const struct sockaddr* , const char* streamid)
{
    // Here we should check if the client requests file transmission type in the streamid. Omitted for simplicity.
    int optval = SRTT_FILE;
    EXPECT_EQ(srt_setsockflag(ns, SRTO_TRANSTYPE, &optval, sizeof optval), SRT_SUCCESS);
    return 0;
}

```
